### ISSUE TITLE
Task tab errors

### ISSUE THREAD

**Issue by kbray1967** [dotproject/dotProject#144]
Installed dotproject. Task tab has lots of errors

ERROR: C:\dotProject-2.2.0\includes\db_adodb.php(67): Error executing: 
SELECT ta.*,pr.project_name, pr.project_id, pr.project_color_identifier,tp.task_pinned FROM (`dotp_tasks` as ta) LEFT JOIN `dotp_projects` AS pr ON pr.project_id = ta.task_project INNER JOIN `dotp_user_tasks` AS ut ON ut.task_id = ta.task_id AND ut.user_id = 1 LEFT JOIN `dotp_user_task_pin` AS tp ON tp.task_id = ta.task_id AND tp.user_id = 1 WHERE (ta.task_percent_complete < 100 OR ta.task_percent_complete IS NULL) AND ta.task_status = 0 AND project_status <> 7 AND project_status != 4 AND task_dynamic != 1 AND ta.task_start_date != '' AND ta.task_start_date != '0000-00-00 00:00:00' GROUP BY ta.task_id ORDER BY ta.task_end_date,task_priority DESC
 (Incorrect DATETIME value: '')
Backtrace:
0 C:\dotProject-2.2.0\includes\db_adodb.php:67 dprint('C:\\dotProject-2.2.0\\includes\\db_adodb.php',67,0,'Error executing: 
SELECT ta.*,pr.project_name, pr.project_id, pr.project_color_identifier,tp.task_pinned FROM (`dotp_tasks` as ta) LEFT JOIN `dotp_projects` AS pr ON pr.project_id = ta.task_project INNER JOIN `dotp_user_tasks` AS ut ON ut.task_id = ta.task_id AND ut.user_id = 1 LEFT JOIN `dotp_user_task_pin` AS tp ON tp.task_id = ta.task_id AND tp.user_id = 1 WHERE (ta.task_percent_complete < 100 OR ta.task_percent_complete IS NULL) AND ta.task_status = 0 AND project_status <> 7 AND project_status != 4 AND task_dynamic != 1 AND ta.task_start_date != \'\' AND ta.task_start_date != \'0000-00-00 00:00:00\' GROUP BY ta.task_id ORDER BY ta.task_end_date,task_priority DESC
 (Incorrect DATETIME value: \'\')')
1 C:\dotProject-2.2.0\includes\db_connect.php:141 db_exec('SELECT ta.*,pr.project_name, pr.project_id, pr.project_color_identifier,tp.task_pinned FROM (`dotp_tasks` as ta) LEFT JOIN `dotp_projects` AS pr ON pr.project_id = ta.task_project INNER JOIN `dotp_user_tasks` AS ut ON ut.task_id = ta.task_id AND ut.user_id = 1 LEFT JOIN `dotp_user_task_pin` AS tp ON tp.task_id = ta.task_id AND tp.user_id = 1 WHERE (ta.task_percent_complete < 100 OR ta.task_percent_complete IS NULL) AND ta.task_status = 0 AND project_status <> 7 AND project_status != 4 AND task_dynamic != 1 AND ta.task_start_date != \'\' AND ta.task_start_date != \'0000-00-00 00:00:00\' GROUP BY ta.task_id ORDER BY ta.task_end_date,task_priority DESC')
2 C:\dotProject-2.2.0\modules\tasks\todo.php:157 db_loadList('SELECT ta.*,pr.project_name, pr.project_id, pr.project_color_identifier,tp.task_pinned FROM (`dotp_tasks` as ta) LEFT JOIN `dotp_projects` AS pr ON pr.project_id = ta.task_project INNER JOIN `dotp_user_tasks` AS ut ON ut.task_id = ta.task_id AND ut.user_id = 1 LEFT JOIN `dotp_user_task_pin` AS tp ON tp.task_id = ta.task_id AND tp.user_id = 1 WHERE (ta.task_percent_complete < 100 OR ta.task_percent_complete IS NULL) AND ta.task_status = 0 AND project_status <> 7 AND project_status != 4 AND task_dynamic != 1 AND ta.task_start_date != \'\' AND ta.task_start_date != \'0000-00-00 00:00:00\' GROUP BY ta.task_id ORDER BY ta.task_end_date,task_priority DESC')
3 C:\dotProject-2.2.0\modules\calendar\vw_day_tasks.php:41 include('C:\\dotProject-2.2.0\\modules\\tasks\\todo.php')
4 C:\dotProject-2.2.0\style\default\overrides.php:95 require('C:\\dotProject-2.2.0\\modules\\calendar\\vw_day_tasks.php')
5 C:\dotProject-2.2.0\modules\calendar\day_view.php:133 show()
6 C:\dotProject-2.2.0\index.php:301 require('C:\\dotProject-2.2.0\\modules\\calendar\\day_view.php')

Warning: count(): Parameter must be an array or an object that implements Countable in C:\dotProject-2.2.0\modules\tasks\todo.php on line 162

Also keeps stating 'Incorrect DATETIME value'

**Comment by ipeevski** [comment]
There was a "fix" in mysql 8.0.16 that doesn't allow `''` dates comparison anymore
I'm guessing you are trying on mysql > 8.0.16.
Can you please try mysql <= 8.0.15 to confirm the issue?

The fix for newer mysql is to remove in the codebase all the references to `..._date != ''` and replace with `..._date is not null` and that also fixes the issue, but it is present in a few places and it might have some side effects

**Comment by kbray1967** [comment]
Wow...well the current version of MySQL is 8.0.22. I can't see reverting to an old version like that. Rather I would expect someone to update DotProject. I did have a look through all the sql and I did not find anywhere 'date !=''. May have to abandon using Dotproject which is unfortunate. I did notice that it has the same structure as Web2Project. I'm guessing the same base code for both as both have the same issue...ugh

**Comment by ipeevski** [comment]
I meant search for `date != ''`

`modules/tasks/todo.php:137`

Change
```
	$q->addWhere("ta.task_start_date != '' AND ta.task_start_date != '0000-00-00 00:00:00'");
```

to
```
	$q->addWhere("ta.task_start_date is not null AND ta.task_start_date != '0000-00-00 00:00:00'");
```

**Comment by kbray1967** [comment]
Yes I found that and a few others. I was able to eliminate 3 errors but could not get the rest. Unfortunately I don't have the time to go through all the  code so I will have to shelve dotproject for now. Maybe when someone releases a new version I will try again.

**Comment by akgreen-777** [comment]
I found and corrected code in 3 places and have had no difficulties since.  the changes are:
todo.php:  line 137  $q->addWhere( "ta.task_start_date is not null AND ta.task_start_date != '0000-00-00 00:00:00'" );
tasks.php:  line 232   . "AND (task_percent_complete < 100 OR task_end_date is null) "
                 line 240   . "AND (task_percent_complete < 100 OR task_end_date is null) "
